feat/ #93 카드 등록/수정 API Refactor

zz106603 opened the pull request:
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] ⚰️ Maintain (keep in good condition or in working)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

<!-- 이슈를 해결한 코드에 대해 설명해주세요. -->

- 새로운 카드 등록, 기존 카드 수정 메서드 분리
- 이미지, 오디오 각 1개씩이기 때문에 Multipart 배열 처리 제거 후, files를 image와 audio로 분리
- update는 필드가 없을 경우 무시, 빈 파일일 경우 기존 파일 삭제, 새로운 파일은 기존 파일 삭제 후 업데이트
- 기존 attachment Bulk 처리 메서드는 유지하고 단일 처리 메서드 추가
- 카드 Service 로직 메서드 분리
- accounId로 권한 있는지 확인하는 로직 추가

### 📝 Checklist

<!-- 테스트 코드 구성이 가능하다면 포함해주세요. -->
<!-- 문서화가 따로 필요 없는 부분이라면 해당 항목을 삭제해주세요. -->

- [x] 🔴 Human eyes (no test)
- [x] 🟡 swagger or Smoke test
- [ ] 🟢 unit test or CI test

### 📸 Log/Screenshot

<!-- 로그나 테스트 스크린샷이 있을 경우 첨부해주세요. -->

🌱 before

🏡 after

zz106603 committed:
#93 카드 등록/수정 API 리팩토링
- 새로운 카드 등록, 기존 카드 수정 메서드 분리
- 이미지, 오디오 각 1개씩이기 때문에 Multipart 배열 처리 제거
- update는 필드가 없을경우 무시, 빈 파일일 경우 기존 파일 삭제, 새로운 파일은 기존 파일 삭제 후 업데이트
- 기존 attachment 벌크 메서드는 유지하고 단일 처리 메서드 추가

# Conflicts:
#	src/main/java/com/edio/studywithcard/attachment/service/S3ServiceImpl.java

zz106603 committed:
feat: 카드 등록/수정 서비스 로직 메서드 분리
- 등록, 수정 각각 메서드 로직 분리
- accountId를 사용하여 등록, 수정 권한 가능한 사용자인지 검사

GuardJo commented on src/main/java/com/edio/common/exception/handler/GlobalExceptionHandler.java:
```diff
@@ -27,6 +28,17 @@ public ResponseEntity<String> handleIllegalArgumentException(IllegalArgumentExce
+    }
+
     /**
      * NoSuchElementException
```
지금 404 예외 처리가 예외는 다르지만 중복 요소라 통합하는게 좋을 듯 합니다

GuardJo replied:
누가 이렇게 한겨 ㅡㅡ

GuardJo commented on src/main/java/com/edio/studywithcard/card/controller/CardApiDoc.java:
```diff
@@ -17,7 +19,8 @@ public interface CardApiDoc {
      */
     @Operation(summary = "Card 등록|수정", description = "Card를 등록하거나 수정합니다.")
-    void createOrUpdateCards(@ModelAttribute CardBulkRequestWrapper cardBulkRequestWrapper);
+    void createOrUpdateCards(@AuthenticationPrincipal CustomUserDetails userDetails,
```
윤환님 해당 인터페이스는 springdoc 용이라 spring의 어노테이션을 쓸 필요 없습니다

GuardJo commented on src/main/java/com/edio/studywithcard/card/model/request/CardBulkRequest.java:
```diff
@@ -16,19 +13,18 @@ public class CardBulkRequest {
                 ", name='" + name + '\'' +
                 ", description='" + description + '\'' +
-                ", files=" + fileNames +
+                ", image=" + (image != null ? image.getOriginalFilename() : "null") +
```
String의 기본 제공 메소드인 valueOf() 메소드를 사용하면 객체가 null값일 경우 "null"을 출력하게 되어 있습니댜
- 굳이 3항연산 쓸 필요 없음
- 근데 어차피 문자열 연산은 내부적으로 String.valueOf() 메소드로 변환되니 null인 객체 그냥 써도 "null"로 출력될거임

GuardJo commented on src/main/java/com/edio/studywithcard/card/service/CardService.java:
```diff
@@ -5,7 +5,7 @@
 
 public interface CardService {
-    void createOrUpdateCards(CardBulkRequestWrapper cardBulkRequestWrapper);
+    void createOrUpdateCards(Long accountId, CardBulkRequestWrapper cardBulkRequestWrapper);
```
가만 생각해보니 createOrUpdate 라는 이름보단 upsert가 더 간지나고 직관적인듯 한데, 어떻게 생각하시나요?

GuardJo commented on src/main/java/com/edio/studywithcard/card/service/CardServiceImpl.java:
```diff
@@ -134,6 +78,58 @@ public void deleteCards(List<Long> request) {
+                    .name(request.getName())
+                    .description(request.getDescription())
+                    .build();
+            cardRepository.save(card);
```
save 하나씩 여러번 쿼리 날리는 것보다 뭉쳐서 한번에 날리는게 더 빠릅니다

zz106603 replied:
> save 하나씩 여러번 쿼리 날리는 것보다 뭉쳐서 한번에 날리는게 더 빠릅니다

현재 save는 객체 리스트에 저장하고 saveAll로 처리해놓았고 attachment같은 경우도 벌크로 처리가 가능할 것 같아서 처리하고 다시 리뷰 요청 드릴게요

zz106603 replied:
attachment, attachmentCardTarget Bulk 처리 적용했습니다

GuardJo commented on src/main/java/com/edio/studywithcard/card/service/CardServiceImpl.java:
```diff
@@ -134,6 +78,58 @@ public void deleteCards(List<Long> request) {
+                            ErrorMessages.NOT_FOUND_ENTITY.format(Deck.class.getSimpleName(), request.getDeckId())
+                    ));
+
+            validateOwnership(accountId, deck);
```
지금 기획 및 시나리오 상 서로 다른 덱의 카드들이 요청 들어올 수 있는지? 만약 없으면 벌크처리 최상단 부분에서 검증 작업 처리하는게 효율적일 듯 합니다
- 지금은 같은 덱의 카드들을 요청 카드 개수만큼 계속 확인 함

GuardJo commented on src/main/java/com/edio/studywithcard/card/service/CardServiceImpl.java:
```diff
@@ -158,4 +154,41 @@ private void processAttachment(MultipartFile file, Card card) throws IOException
+
+    // 신규 첨부파일 저장 (등록 시)
+    private void processNewAttachment(MultipartFile file, Card card) throws IOException {
+        if (file != null && !file.isEmpty()) {
```
spring에서 MultipartFile은 파일 입력 없을 시에도 메타데이터 값이 있기에 null은 입력 안될 겁니다
- 빈 데이터 여부만 체크해줘도 될 듯 함

zz106603 replied:
> spring에서 MultipartFile은 파일 입력 없을 시에도 메타데이터 값이 있기에 null은 입력 안될 겁니다
> 
> * 빈 데이터 여부만 체크해줘도 될 듯 함

1. 파라미터가 아예 안들어오는 경우 (파일 변경이 없을 경우)
2. 파라미터는 들어오지만 빈 파일의 경우 (기존 파일 삭제할 경우)
3. 파라미터에 파일이 넘어오는경우 (파일 추가)

이렇게 3개로 처리해야되기때문에 null처리를 안해주면 NullPointException이 발생할 것 같습니다

zz106603 committed:
feat: card 생성/수정 API 리팩토링
- card 생성은 벌크로 한번에 처리(saveAll)
- card 수정은 dirty check로 update 유지
- 첨부파일은 일단 각 실행되도록 하고 추후에 벌크 작업으로 리팩토링 예정

zz106603 committed:
feat: Card 삭제 개선(deckId 포함해서 소유권 검증 추가)

zz106603 commented:
conflict 발생해서 main 코드 가져와서 충돌 해결하느라 SonarQube 리팩토링 코드도 들어갔네요..

zz106603 committed:
feat: api doc 주석 내용 수정

jsjouo249 commented on src/main/java/com/edio/studywithcard/card/service/CardServiceImpl.java:
```diff
@@ -158,4 +175,41 @@ private void processAttachment(MultipartFile file, Card card) throws IOException
+
+        if (file.isEmpty()) {
+            // 빈 파일이면 기존 파일 삭제
+            if (fileKey != null) {
```
199line의 if와 204line의 if가 동일한 것 같은데,

197line의 if문 밖으로 꺼내면, 한 번에 확인할 수 있을 것 같습니다.

zz106603 replied:
> 199line의 if와 204line의 if가 동일한 것 같은데,
> 
> 197line의 if문 밖으로 꺼내면, 한 번에 확인할 수 있을 것 같습니다.

좋은 지적입니다

jsjouo249 commented on src/main/java/com/edio/studywithcard/card/service/CardServiceImpl.java:
```diff
@@ -42,98 +43,114 @@ public class CardServiceImpl implements CardService {
+
+        // 첫 요청에서 Deck ID를 가져옴
+        Long deckId = cardBulkRequestWrapper.getRequests().get(0).getDeckId();
+        Deck deck = deckRepository.findById(deckId).orElseThrow(() -> new EntityNotFoundException(ErrorMessages.NOT_FOUND_ENTITY.format(Deck.class.getSimpleName(), deckId)));
```
service에서 findById를 호출할 때 마다, entityNotFoundException을 처리하지말고,
repository에서 한번에 처리하는건 어떠실까요??

```
Optional<Card> findByIdAndIsDeletedFalse(Long id) {
    Optional<Card> card = this.findByIdAndIsDeletedFalse(id);
    if (card != null) {
        return card;
    } else {
        throw new NotFoundException();
    }
};
```

모두의 의견을 받고 싶습니다.

zz106603 replied:
레포지토리에서 처리한다는게 무슨말인지 이해를 못했는데 repository에서 직접 예외를 던진다는 말인가여?

zz106603 replied:
repository는 DAO라서 비즈니스 로직이 포함되면 안된다고 알고있슴다 만약에 중복 코드를 줄이려면 다른 방법을 써야할거같슴다

jsjouo249 replied:
넵넵

저렇게 사용하는게 jpa를 잘 사용하는건지는 알지못하지만,
제가 임의로 작성한 코드가 repository 파일 예시인데,
저렇게 repo에서 직접 에러를 반환하면,
각 서비스에서 repo호출할 때 마다 에러 검증 안하고, 함수만 호출하면 되니까 더 편하지 않을까 해서 입니다.

get처럼 가져와야하는데, 데이터가 없어서 에러를 내는 함수들에만 넣으면 호출하는 부분도 편해지고, 에러도 한 곳에서 처리하니 더 낫지 않을까 합니다.

jsjouo249 replied:
비즈니스 로직은 서비스에 있는게 맞는데,
단순히 데이터의 존재 여부에 대한 예외 던지는 것도 비즈니스 로직에 속하는걸로 취급하는 것이군요.

의견 감사합니다

zz106603 replied:
repository에서 예외 처리하게되면 비즈니스 계층에서 처리해야 할 예외를 데이터 계층에서 다루고있기 때문에 단일책임원칙에 위배된다고 하네영

GuardJo replied:
QueryDsl 이나 Criteria처럼 DAO를 wrapping해서 일부 로직을 추가하여 사용하는 경우들도 존재합니다.

다만, 해당 요소들은 주로 복잡한 검색이나 조회 등에 사용하는 경우가 많으며, 지금과 같은 단순 1개의 Table 내 인스턴스는 굳이 해당 요소들을 사용할 필요는 없어보입니다.

성ㅈ님이 말씀하신 사항도 공감되나, 특정 Entity가 조회 되지 않을 경우 무조건적으로 EntitiyNotFoundException을 던지게 되면, 지금같은 로직에선 문제 없으나, Entity가 없을 경우 신규 데이터를 생성하는 등의 시나리오에는 추가 예외 처리가 결국 필요해지게 됩니다.

EntityNotFoundException의 경우, Unchecked Exception이기에 성ㅈ님이 말씀하신 요소로 구성될 경우, 사용자들이 해당 인터페이스 내부 로직을 알고 있어야 하는 점이 존재합니다.
- Checked Exception으로 강제할 경우에는 괜찮긴 함

근데 현 모듈 내에 Entity들 조회 시 별도 시나리오가 없을 경우에는 성ㅈ님이 말씀하신 내용 반영도 좋을 듯 합니다.
다만 해당 의견을 소스에 반영한다면 DAO 내 직접 구성하는건 윾난님 의견대로 해당 객체에 주는 부담이 많아지며, 제 의견으로는 BaseEnttiy를 기준으로 AOP를 활용하여 처리하면 어떨까 싶습니다

zz106603 replied:
AOP를 활용해서 findById로 시작하는 것들은 공통으로 처리할 수 있도록 적용했습니다

zz106603 commented:
카드 등록/수정 테스트 코드 추가했습니다

zz106603 committed:
feat: CardServiceTests 테스트 로직 추가

zz106603 committed:
feat: 소유권 검증 로직 개선(JPQL)으로 조회 성능 최적화

zz106603 commented on src/main/java/com/edio/studywithcard/deck/repository/DeckRepository.java:
```diff
@@ -2,9 +2,15 @@
+
+    // 소유권 검증을 위한 JPQL
+    @Query("SELECT f.accountId FROM Deck d JOIN d.folder f WHERE d.id = :deckId")
+    Long findAccountIdByDeckId(@Param("deckId") Long deckId);
```
- 이전 코드에서는 Deck엔티티 조회 -> 연관 폴더 조회 -> 사용자 accountId 조회 순으로 기존 accountId를 가져왔는데 이렇게 엔티티를 여러번 호출하는건 비효율적이라고 생각
- 어짜피 엔티티 조작같은거 필요없고 accountId만 필요하기 때문에 쿼리는 한번만 날리는 JPQL방식으로 적용했는데 피드백 부탁드립니다.

GuardJo replied:
Card Entity를 신규 등록 할 경우엔 Deck을 조회하고, Card Entity를 수정하는 경우엔 Card를 조회해오고 있는데, 해당 부분에서 owner여부 체크를 진행하면 해당 쿼리 호출 없이도 가능하지 않나요?

해당 메소드를 사용할 경우 owner 검증 (쿼리 1회 요청) -> 신규저장/수정 (쿼리 2회(덱 or 카드 조회) 요청) -> 트랜잭션 종료 (update 쿼리 1회)

해당 메소드를 사용하지 않고 저장/수정 시점에 owner 검증할 경우 신규저장/수정 (쿼리 2회(덱 or 카드 조회) 요청) -> 트랜잭션 종료 (update 쿼리 1회)

위와 같이 오히려 안쓰는게 DB I/O를 줄일 수 있습니다.
또한 query 어노테이션 사용 시에는 JPA의 Persistence Context에서 해당 Entity를 관리하지 않고 호출하기 때문에 Deck Entity가 해당 트랜잭션 내에서 이미 올라와 있더라도 무조건 호출하게 됩니다.

GuardJo replied:
말씀하신 Deck -> Folder 체이닝은 실제 DB 쿼리 호출이 아닌 Persistence Context 상에 올라온 메모리 내에서 처리되고 있는 점 참고 부탁드립니다. (이미 메모리상에 올라와 있는 애라 딱히 비효율적이 아님)
- 만약 LazyLoading으로 설정했을 경우엔 각 요소 접근 시마다 쿼리가 날라갔겠지만, 현재 해당 요소들은 EAGER라 상관 없음

zz106603 replied:
카드 생성 시점에는 조회된 Deck으로 accountId 찾기, 
카드 수정 시점에는 조회된 Card로 accountId 찾기로 수정했습니다.

카드 삭제인 경우에는 첫 번째 index의 Card로 accountId 찾기로 적용이요

zz106603 committed:
feat: 소유권 검증 수정으로 테스트 코드 refactor

zz106603 committed:
충돌 로컬 변경사항 저장

zz106603 committed:
feat: Attachment Bulk 처리 테스트 로직 수정

zz106603 committed:
feat: 카드 수정시 attachment, attachmentCardTarget Bulk 처리
- card, attachment 테스트 로직 수정
- 카드 수정시 파일(삭제, 유지, 수정) Bulk 처리 적용

zz106603 committed:
feat: CardService 공통 로직 분리 및 코드 리팩토링

zz106603 commented:
upsert에서 Card, Attachment, AttachmentCardTarget을 Bulk로 처리하도록 전체적으로 수정했으니 다시 검토 요청드립니다.

zz106603 committed:
feat: card 생성 null 검증 수정

GuardJo commented on src/main/java/com/edio/studywithcard/attachment/service/AttachmentService.java:
```diff
@@ -12,12 +13,15 @@ public interface AttachmentService {
     void deleteAllAttachments(List<String> fileKeys);
+
+    // 첨부 파일 삭제(단일)
+    void deleteAttachment(String fileKey);
```
위에 저장도 그렇고 여기 삭제도 그렇고 굳이 다건 삭제가 있는데 단일이 있을 필요가 있나요?
- 다건 처리 메소드 인자에 하나만 넣으면 되잖아요

zz106603 replied:
> 위에 저장도 그렇고 여기 삭제도 그렇고 굳이 다건 삭제가 있는데 단일이 있을 필요가 있나요?
> 
> * 다건 처리 메소드 인자에 하나만 넣으면 되잖아요

보니까 이미 deleteAttachment는 안쓰고있어서 제거요

GuardJo commented on src/main/java/com/edio/studywithcard/card/model/request/CardBulkRequest.java:
```diff
@@ -16,19 +13,18 @@ public class CardBulkRequest {
                 ", description='" + description + '\'' +
-                ", files=" + fileNames +
+                ", image='" + image + '\'' +
+                ", audio='" + audio + '\'' +
```
MultipartFile toString() overriding 되어 있나요?

zz106603 replied:
> MultipartFile toString() overriding 되어 있나요?

일단 기본적으로 MultipartFile 자체로는 toString이 없어서 object.toString() 형태로 반환되고 재정의한거고
image나 audio에 의미없는 클래스이름이 나오기때문에 null 처리 포함한 .getOriginalFilename() 적용할게여

zz106603 replied:
그리고 위 리뷰에서 valueOf()으로 자동 전환이라 null처리 안해도된다고 하셨는데 
.getOriginalFilename()을 사용하게 되면 .valueOf() 처리 전에 image에서 getOriginalFilename을 하기 때문에 NPE가 발생해서 null 처리 포함해서 적용이여

GuardJo commented on src/main/java/com/edio/studywithcard/card/service/CardServiceImpl.java:
```diff
@@ -42,120 +45,177 @@ public class CardServiceImpl implements CardService {
+        // 소유권 검증
+        Long deckId = cardBulkRequestWrapper.getRequests().get(0).getDeckId();
+        Long ownerId = deckRepository.findAccountIdByDeckId(deckId);
+        validateOwnership(accountId, ownerId);
```
코멘트 달았던 부분인데, 이 부분을 Deck이나 Card 조회 시점에서 처리하면 쿼리 더 호출 안해도 됩니다

GuardJo commented on src/main/java/com/edio/studywithcard/card/service/CardServiceImpl.java:
```diff
@@ -42,120 +45,177 @@ public class CardServiceImpl implements CardService {
         }
+
+        // 신규 카드 저장
+        if (!newCards.isEmpty()) {
```
이거랑 밑에 attachment랑 List 빈 여부 체크 안해도 됩니다.
- saveAll() 구현체 로직을 보면 어차피 iterator로 순회 도는데, 데이터 없으면 그냥 넘어가니 체크 불필요

GuardJo commented on src/main/java/com/edio/studywithcard/card/service/CardServiceImpl.java:
```diff
@@ -42,120 +45,177 @@ public class CardServiceImpl implements CardService {
+                .map(AttachmentBulkData::getOldFileKey)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+        if (!oldFileKeys.isEmpty()) {
```
얘네도 굳이 빈 거 체크할 필요 없습니다
- 이유는 위에 언급함

zz106603 committed:
feat: 불필요한 if문 제거

zz106603 committed:
feat: EntityNotFound Exception AOP로 처리

zz106603 committed:
feat: 소유권 검증 및 테스트 로직 수정

zz106603 committed:
feat: 단일 attachment 삭제 메서드 제거

jsjouo249 commented on src/main/java/com/edio/common/aop/EntityNotFoundAspect.java:
```diff
@@ -0,0 +1,35 @@
+@Slf4j
+public class EntityNotFoundAspect {
+
+    @Around("execution(* org.springframework.data.jpa.repository.JpaRepository+.*(..)) && " +
```
이건 제가 잘 몰라서 궁금증인데
notfound 체크인데, @AfterRunning이 아닌 @around로 하신 이유는 어떤 이유 때문이신건가요??

zz106603 replied:
> 이건 제가 잘 몰라서 궁금증인데 notfound 체크인데, @AfterRunning이 아닌 @around로 하신 이유는 어떤 이유 때문이신건가요??

혹시 AfterRunning이 아니고 AfterReturning을 물어본건가요 afterRunning이라는건 없어서.. 

일단 around랑 afterReturning 모두 Repository에서 Optional 반환을 한다면 optional.isEmpty()로 검사해서 예외를 던질 수 있습니다. 현재 Repository에서 다 Optional로 되어있기 때문에 around, afterReturning 둘다 사용해도 무방하지만 추후에 내부(findById ..를 호출하는 부분)에서 따로 예외를 처리해야 하는 상황이 생기면 afterReturning은 동작하지 않기때문에(메서드가 성공해야 aop 동작)  확장성 측면에서 around가 더 맞다고 생각함다

jsjouo249 replied:
아 넵넵
aftrerReturning이였습니다.

> 추후에 내부(findById ..를 호출하는 부분)에서 따로 예외를 처리해야 하는 상황이 생기면

이건 XXXRepository.java 내부에서 처리 말씀하시는거죠??
넵넵 답변 감삼다

zz106603 replied:
> 아 넵넵 aftrerReturning이였습니다.
> 
> > 추후에 내부(findById ..를 호출하는 부분)에서 따로 예외를 처리해야 하는 상황이 생기면
> 
> 이건 XXXRepository.java 내부에서 처리 말씀하시는거죠?? 넵넵 답변 감삼다

아 말을 제가 잘못했네요. 확장성 측면이 아니고 반대네요 afterRunning을 사용하게 되면 service나 이런데서 cardRepository.findById().orElseThrow()를 사용하게 되면 aop로 넘어오지 않게 되서 무조건 findById로 시작하는건 aop에서 처리하도록 하기 위해서 around를 사용했슴다

zz106603 committed:
feat: bulkRequest toString() 수정

GuardJo reviewed (approved):
👍
